keps/sig-cli/kustomize-secret-generator-plugins.md

```diff
@@ -0,0 +1,174 @@
+---
+title: Kustomize Secret Generator Plugins
+authors:
+  - "@sethpollack"
+owning-sig: sig-cli
+participating-sigs:
+  - sig-apps
+  - sig-architecture
+reviewers:
+  - "@monopole"
+  - "@Liujingfang1"
+approvers:
+  - "@monopole"
+  - "@Liujingfang1"
+  - "@pwittrock"
+editor: "@sethpollack"
+creation-date: 2019-02-04
+last-updated: 2019-02-04
+status: provisional
+---
+
+# Kustomize Secret Generator Plugins
+
+## Table of Contents
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+Kustomize wants to generate Kubernetes Secret objects from general key-value pairs where the value is supposed to be a secret. Currently Kustomize only supports reading secret values from local files which raises security concerns about file lifetime and access. Reading secret values from the execution of arbitrary "commands" in a kustomization.yaml file introduces concerns in a world where kustomization files can be used directly from the internet when a user runs `kustomize build`.
+
+This proposal describes obtaining secret values safely using golang [plugins](https://golang.org/pkg/plugin/).
+
+## Motivation
+
+Not having a way to integrate `SecretGenerator` with secret management tools requires hacky workarounds.
+
+### Goals
+
+- Give users a way to intergrate `SecretGenerator` with secret management tools in a safe way.
+
+### Non-Goals
+
+- Kustomize will not handle plugin installation/management.
+
+## Proposal
+
+In the proposed design `SecretGenerator` would load golang plugins from `~/.config/kustomization_plugins`. This would limit the scope of what kustomize can execute to the plugins on the users machine. Also using golang plugins forces a strict interface with well-defined responsibilities.
+
+The current API looks like this:
+
+```
+secretGenerator:
+- name: app-tls
+  files:
+  - secret/tls.cert
+  - secret/tls.key
+  type: "kubernetes.io/tls"
+
+- name: env_file_secret
+  env: env.txt
+  type: Opaque
+
+- name: myJavaServerEnvVars
+  literals:
+  - JAVA_HOME=/opt/java/jdk
+  - JAVA_TOOL_OPTIONS=-agentlib:hprof
+```
+
+The proposed API would look like this:
+
+```secretGenerator:
+- name: env-example
+  goplugins:
+  - name: Env
+    args:
+    - EXAMPLE_ENV
+    - OTHER_EXAMPLE_ENV
+  type: Opaque
+```
+
```

Comment:
How about kustomization file changes from [this](https://github.com/kubernetes-sigs/kustomize/blob/master/pkg/target/configmaps_test.go#L25):
```
secretGenerator:
  - name: bob
    literals:
    - fruit=apple
    - vegetable=broccoli
    env: foo.env
    files:
    - passphrase=phrase.dat
    - forces.txt
  - name: sally
     ...
```
to the following:
```
  secretGenerator:
  - name: bob
    sources:
    - strategy: literal
      args:
      - fruit=apple
      - vegetable=broccoli
    - strategy: env
      args:
      - foo.env
    - strategy: files
      args:
      - passphrase=phrase.dat
      - forces.txt
    - strategy: goplugin
      name: awesome
      args:
      - foo
      - bar
    - stratagy: goplugin
      name: usegrpc
      args:
      - foo
      - bar
    - strategy: kubectlplugin
      name: whatever
      args:
      - foo
      - bar
  - name: sally
     ...
```
The above reformulates the existing _name=value_ pair generators (`literals`, `files` and `env`) as builtin _strategies_.

 This stanza suggests two new strategies, `goplugin` and `kubectlplugin`.

There are two instances of `goplugin` called `awesome` and `usegrpc`, and one instance of `kubectlplugin` called `whatever`.

Then a strategy instance graduates to builtin by making its name a new reserved word in the strategy field, e.g. this:
```
    - strategy: goplugin
      name: awesome
      args:
      - foo
      - bar
```
could graduate to
```
    - strategy: awesome
      args:
      - foo
      - bar
```

This way we can add goplugins, open the door to other kinds of plugins, and allow a graduation strategy to become a builtin like the existing `literals`, `files` and `env`.

We could deprecate the existing `DataSources` format, and modify the existing `fix` command to make this change trivial.  We've done this before.

Also, if we do the impl in `GeneratorArgs`, you get all these benefits in `ConfigMap` as well as `Secret`.

Reply:
Yeah, thats basically what I had in mind. I just went with `plugins` instead of `sources` and `type` instead of `strategy`.

For the first phase I ported `DataSources` over to a new `BuiltIn` plugin type and then in another PR we can deprecate `DataSources` and add a `fix` command. 

We can also make `kustomize add secret/configmap` more generic with `--plugin-type`, `--plugin-name`, and `--args` which would work for all plugins.

```
  secretGenerator:
  - name: bob
    plugins:
    - type: BuiltIn
      name: literal
      args:
      - fruit=apple
      - vegetable=broccoli
    - type: BuiltIn  
      name: env
      args:
      - foo.env
    - type: BuiltIn  
      name: files
      args:
      - passphrase=phrase.dat
      - forces.txt
    - type: Go
      name: awesome
      args:
      - foo
      - bar
    - type: Go
      name: usegrpc
      args:
      - foo
      - bar
    - type: Kubectl
      name: whatever
      args:
      - foo
      - bar
```
I already updated the PR, but I can rename those fields if you would like.

Reply:
Thanks!

W/r to the word _plugin_ - i want to bury it a little deeper.

Because everything that makes sense about this approach to KV pair generation makes even more sense for transformers - a completely different "plugin" thing.

So we should eschew a highly visible package, interface, or type called called simply `plugin` or `Plugin`.  Better to call the interface `KvSourcePlugin`, and put the code under, say,
`k8sdeps/kv/plugin` instead of in `pkg/plugin`.

Also suggest we likewise push the word `plugin` a little deeper in the kustomization file - see below.

#### Nice stuff in PR kubernetes-sigs/kustomize#760

Modulo the use of the bare word plugin and pluginFactory, the overall approach is great.
 - there's a `Plugin` interface with a `Get` method that returns kv pairs (prefer a rename to `KvSourcePlugin`).
 - it's clear that there are two types of plugins, `Go` and `BuiltIn`, and a factory for making instances.
 - the legacy sources - `literals`, `env`, and `files` - are converted to `Builtin` form
 - there's a clear path (clear to me anyway) for making _exec_ style plugins, or a _kubectl plugin_ runner.
 - Would prefer some changes to field names in the kustomization file:

Instead of
```
 plugins:
 - type: BuiltIn
   name: env
   ...
 - type: BuiltIn
   name: literals
   ...
 - type: Kubectl
   name: whatever
   ...
 - type: Go
   name: awesome
   ...
```
suggest:
```
 kvSources:
  - name: env
    ...
  - name: literals
    ...
  - plugin: kubectl
    name: whatever
    ...
  - plugin: go
    name: awesome
    ...
```

i.e. i prefer pushing the word plugin down a level in the kustomization file, and using the reserved words `builtin, go, kubectl` instead of `BuiltIn, Go, Kubectl`, and i'd allow the `plugin` field to default to `builtin` (so it can be optional).

Reply:
Ok sounds like a plan, I'll update the PR.

Reply:
@monopole Can we go with `type` instead of  `plugin`? It makes it seem like it is a single plugin vs an instance of a plugin type.

Reply:
sure.

Reply:
1
```
plugin: kubectl
plugin: go
```

2
```
type: kubectlPlugin
type: goPlugin
```

3
```
type: kubectl
type: go
```

4
```
pluginType: kubectl
pluginType: go
```


3 is brief, but 4 is self-explanatory.

Reply:
Yeah, I'm ok with 3 or 4